### PR TITLE
Ci/197 docker images

### PR DESCRIPTION
**Pull request by Uruael**
## Overview
  * Status: 
    - [x] ✅ Ready for review
    - [ ] ⚠️ Hold
    - [ ] 🚧 WIP
    
  * Type: 
    - [ ] 🛠 Build
    - [x] 👷‍ CI
    - [ ] 📚 Docs
    - [ ] ✨ Feat
    - [ ] 🐛 Fix
    - [ ] 🚀 Perf
    - [ ] 📦 Refactor
    - [ ] 💎 Style    
    - [ ] 🧪 Test

    

## Description

Created reusable workflow for building and publishing docker images. 
- Setup: Checkout necessary repositories, prepare certs, login to a container registry, setup qemu, and builder
- `Docker meta` step prepares metadata for the image. 
  - Tag latest will be added automatically for releases on main. 
  - Prepared versioning for semantic releases (`type=semver`)
  - Every image can be identified by tag `build-{branch-name}-{github.run_id}-{commit-sha}`
- `Build and push`: Build docker image for desired architecture with buildx and push it to the registry. 

## Motivation and Context

https://github.com/Modino-io/modino/issues/197

## How Has This Been Tested?

Tested via Github Actions.

## Related Issues

https://github.com/Modino-io/modino/issues/197

## Checklist:

- [ ] My change requires a change to the documentation (for example ADRs/README).
- [ ] I have updated the documentation accordingly.
- [ ] I have added/updated tests to cover my changes.

### CODE REVIEW

File: .github/workflows/container-reusable.yaml [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+          required: true
+        REGISTRY_LOGIN:
+          required: true
+        REGISTRY_PASSWORD:
+          required: true
+        GITLAB_TOKEN:
+          required: true
+
+jobs:
+  publish-packages:
```
Comment (KrzeQ) [inline .github/workflows/container-reusable.yaml:39]: Publish images sounds better

File: .github/workflows/container-reusable.yaml [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+      inputs:
+        CALLER_REPOSITORY:
+          required: true
+          type: string
+        CALLER_REPOSITORY_VERSION:
+          required: true
+          type: string
+        MODINO_DEVOPS_VERSION:
+          required: true
+          type: string
```
Comment (KrzeQ) [inline .github/workflows/container-reusable.yaml:14]: Maybe BRANCH instead of VERSION?

File: .github/workflows/container-reusable.yaml [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+          required: true
+        GITLAB_TOKEN:
+          required: true
+
+jobs:
+  publish-packages:
+    runs-on: ubuntu-20.04
+    permissions:
+      contents: write
+      packages: write
```
Comment (KrzeQ) [inline .github/workflows/container-reusable.yaml:43]: Do `contents` need to have write permissions or read would be sufficient?

File: .github/workflows/container-reusable.yaml [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    steps:
+      - name: Checkout Modino-io/modino-devops repository
+        uses: actions/checkout@v2
+        with:
+          repository: Modino-io/modino-devops
+          path: modino-devops
+          ref: ${{ inputs.MODINO_DEVOPS_VERSION }}
```
Comment (KrzeQ) [inline .github/workflows/container-reusable.yaml:50]: Same as above with BRANCH instead of VERSION

**Review by KrzeQ — COMMENTED** [review]
Overall looks good, need to test it. Some questions below

File: .github/workflows/container-reusable.yaml [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        env:
+          GITLAB_TOKEN: ${{ secrets.GITLAB_TOKEN }}
+          CALLER_REPOSITORY: ${{ inputs.CALLER_REPOSITORY }}
+
+      - name: Login to Docker Hub
+        uses: docker/login-action@v2
+        with:
+          registry: ${{ inputs.REGISTRY }}
+          username: ${{ secrets.REGISTRY_LOGIN }}
+          password: ${{ secrets.REGISTRY_PASSWORD }}
```
[the remark this answers is not in the document]
Comment (Uruael) [reply .github/workflows/container-reusable.yaml:73]: You can log in to whatever repo you pass in inputs. The name can be `Login to Docker Registry`

File: .github/workflows/container-reusable.yaml (hunk 1 above)
Comment (Uruael) [reply .github/workflows/container-reusable.yaml:39]: 👍

File: .github/workflows/container-reusable.yaml (hunk 2 above)
Comment (Uruael) [reply .github/workflows/container-reusable.yaml:14]: This input can be deleted and we can reference main instead

File: .github/workflows/container-reusable.yaml (hunk 4 above)
Comment (Uruael) [reply .github/workflows/container-reusable.yaml:50]: Same as above, we can reference main

File: .github/workflows/container-reusable.yaml (hunk 3 above)
Comment (Uruael) [reply .github/workflows/container-reusable.yaml:43]: 👍

**Review by KrzeQ — DISMISSED** [review]
👍

**Review by KrzeQ — APPROVED** [review]

**Pull request merged**